torch/hub.py

```diff
@@ -119,29 +121,82 @@ def _read_url(url):
         return r.read().decode(r.headers.get_content_charset('utf-8'))
 
 
-def _validate_not_a_forked_repo(repo_owner, repo_name, branch):
-    # Use urlopen to avoid depending on local git.
-    headers = {'Accept': 'application/vnd.github.v3+json'}
-    token = os.environ.get(ENV_GITHUB_TOKEN)
-    if token is not None:
-        headers['Authorization'] = f'token {token}'
-    for url_prefix in (
-            f'https://api.github.com/repos/{repo_owner}/{repo_name}/branches',
-            f'https://api.github.com/repos/{repo_owner}/{repo_name}/tags'):
-        page = 0
-        while True:
-            page += 1
-            url = f'{url_prefix}?per_page=100&page={page}'
-            response = json.loads(_read_url(Request(url, headers=headers)))
-            # Empty response means no more data to process
-            if not response:
-                break
-            for br in response:
-                if br['name'] == branch or br['commit']['sha'].startswith(branch):
-                    return
+def _validate_branch(repo_owner, repo_name, branch):
+    # Here we try to make sure that the branch isn't a potentially malicious commit.
+    # This is important because in GitHub the download URL
+    # f'https://github.com/{repo_owner}/{repo_name}/archive/{commit_hash}.zip'
+    # may actually exist *even if* commit_hash is from a different repo_owner / fork.
+    # For example this URL exists:
+    # https://github.com/pytorch/vision/archive/8949c7011facf7801fdf077cc3e4ecd8f0940c7e.zip
+    # even though the commit hash doesn't come from pytorch/vision. It actually comes from a fork
+    # https://github.com/NicolasHug/vision/commit/8949c7011facf7801fdf077cc3e4ecd8f0940c7e
+    # So we want avoid downloading code that can come from a potentionally malicious fork.
+
+    try:
+        # The above issue only exists for commit hashes. So if we know that ``branch``
+        # isn't a commit hash, we can return early. This should avoid many GitHub API calls.
+        # If we can't convert branch to a hex int we know for sure it's not a commit hash
+        int(branch, 16)
+    except ValueError:
+        return
+    # Note: here it's still possible that the branch param corresponds to a branch name
+    # or a tag name, so we need to check for those as well.
+
+    def branch_belongs_to_repo():
```

Comment:
Is there a strong reason why this has to be inner function?

I do not oppose much but I get the impression that if we extract these, the `_validate_branch` function only expresses the logical flow, and it's very readable

Reply:
I agree extracting it would make it simpler. Done!